### PR TITLE
Route to endpoint instance

### PR DESCRIPTION
**Pull request by timbussmann**
Connects to Particular/V6Launch#68

Adds the possibility to route to a specific endpoint instance besides of a logical endpoint or a physical endpoint. This was already possible when defining a dynamic rule, but not on the static API. This can be useful for transports providing extensions to EndpointInstance like `AtMachine` for MSMQ or a Schema specific extension for SQL (currently they would have to hardcode an address string instead).

new options:
* `unicastRoutingTable.RouteToInstance(new EndpointInstance("sales", "1").AtMachine("somewhere"));`
* `publishers.Add(new EndpointInstance("sales", "1").AtMachine("somewhere"));`

also refactored the UnicastRouter tests a bit and split them into send/publish specific tests.

Note: This only affects the advanced API and isn't accessible on the endpoint config routing extensions.

@DavidBoike @mauroservienti @Scooletz @Particular/nservicebus-maintainers please review
(/cc @MarcinHoppe )

### CODE REVIEW

**Comment by bording** [comment]
LGTM :+1:

File: src/NServiceBus.Core.Tests/Routing/UnicastPublisherRouterTests.cs [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+            var sales = "Sales";
+            var shipping = "Shipping";
+            subscriptionStorage.Subscribers.Add(new Subscriber("sales1", sales));
+            subscriptionStorage.Subscribers.Add(new Subscriber("sales2", sales));
+            subscriptionStorage.Subscribers.Add(new Subscriber("shipping1", shipping));
+            subscriptionStorage.Subscribers.Add(new Subscriber("shipping2", shipping));
+
+            var routes = (await router.Route(typeof(Event), new DistributionPolicy(), new ContextBag())).ToArray();
+
+            Assert.AreEqual(2, routes.Length);
```
Comment (danielmarbach) [inline src/NServiceBus.Core.Tests/Routing/UnicastPublisherRouterTests.cs]: Can we move the assert down?

File: src/NServiceBus.Core.Tests/Routing/UnicastPublisherRouterTests.cs [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+            var destinations = routes.Select(ExtractDestination).ToList();
+            Assert.Contains("sales1", destinations);
+            Assert.Contains("shipping1", destinations);
+        }
+
+        [Test]
+        public async Task Should_not_send_multiple_copies_of_message_to_one_physical_destination()
```
Comment (danielmarbach) [inline src/NServiceBus.Core.Tests/Routing/UnicastPublisherRouterTests.cs]: Above you use should_route and here should_no_send (below test as well) although you still actually only get routes. Should we avoid the term send?

File: src/NServiceBus.Core.Tests/Routing/UnicastSendRouterTests.cs [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+        [Test]
+        public async Task When_routing_command_to_logical_endpoint_without_configured_instances_should_route_to_a_single_destination()
+        {
+            var logicalEndpointName = "Sales";
+            routingTable.RouteToEndpoint(typeof(Command), logicalEndpointName);
+
+            var routes = await router.Route(typeof(Command), new DistributionPolicy(), new ContextBag());
+
+            Assert.AreEqual(1, routes.Count());
```
Comment (danielmarbach) [inline src/NServiceBus.Core.Tests/Routing/UnicastSendRouterTests.cs]: Move assert

File: src/NServiceBus.Core.Tests/Routing/UnicastSendRouterTests.cs [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+            var routes = (await router.Route(typeof(Command), new DistributionPolicy(), new ContextBag())).ToArray();
+
+            Assert.AreEqual(2, routes.Length);
+            Assert.AreEqual("Sales-1", ExtractDestination(routes[0]));
+            Assert.AreEqual("Shipping-1", ExtractDestination(routes[1]));
+        }
+
+        [Test]
+        public async Task Should_not_send_multiple_copies_of_message_to_one_physical_destination()
```
Comment (danielmarbach) [inline src/NServiceBus.Core.Tests/Routing/UnicastSendRouterTests.cs]: Same comment as in other test

**Comment by danielmarbach** [comment]
Made a few comments

File: src/NServiceBus.Core.Tests/Routing/UnicastPublisherRouterTests.cs (hunk 2 above)
Comment (timbussmann) [inline src/NServiceBus.Core.Tests/Routing/UnicastPublisherRouterTests.cs]: makes sense to me :+1:

File: src/NServiceBus.Core.Tests/Routing/UnicastSendRouterTests.cs (hunk 3 above)
Comment (timbussmann) [inline src/NServiceBus.Core.Tests/Routing/UnicastSendRouterTests.cs]: that doesn't really make sense here? I inlined the destination variable though (on the next line).

**Comment by timbussmann** [comment]
@danielmarbach addressed your comments

**Pull request merged**